File: src/content/docs/topics/cs/coding-problems/arrays-and-hashing/303-range-sum-query-immutable-approach2.py
class NumArray:
    def __init__(self, nums: list[int]):
        self.prefix = [0] * (len(nums) + 1)
        for i, v in enumerate(nums):
            self.prefix[i + 1] = self.prefix[i] + v

    def sum_range(self, left: int, right: int) -> int:
        return self.prefix[right + 1] - self.prefix[left]

na = NumArray([-2, 0, 3, -5, 2, -1])
assert na.sum_range(0, 2) == 1
assert na.sum_range(2, 5) == -1
assert na.sum_range(0, 5) == -3
na2 = NumArray([5])
assert na2.sum_range(0, 0) == 5
na3 = NumArray([-1, -2, -3])
assert na3.sum_range(0, 2) == -6
assert na3.sum_range(1, 2) == -5
print("all tests pass")

File: src/content/docs/topics/cs/coding-problems/arrays-and-hashing/303-range-sum-query-immutable-approach2.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

class NumArray {
    private prefix: number[];

    constructor(nums: number[]) {
        this.prefix = new Array(nums.length + 1).fill(0);   // L1: O(n) allocate
        for (let i = 0; i < nums.length; i++)
            this.prefix[i + 1] = this.prefix[i] + nums[i]; // L3: O(1) per step
    }

    sumRange(left: number, right: number): number {
        return this.prefix[right + 1] - this.prefix[left];  // L4: O(1)
    }
}

const na = new NumArray([-2, 0, 3, -5, 2, -1]);
assert(na.sumRange(0, 2) === 1);
assert(na.sumRange(2, 5) === -1);
assert(na.sumRange(0, 5) === -3);
const na2 = new NumArray([5]);
assert(na2.sumRange(0, 0) === 5);
const na3 = new NumArray([-1, -2, -3]);
assert(na3.sumRange(0, 2) === -6);
assert(na3.sumRange(1, 2) === -5);
console.log("all tests pass");
